larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,464 @@
+    m_useTrainingMode(false),
+    m_selectNuanceCode(false),
+    m_minPurity(0.9f),
+    m_minCompleteness(0.9f),
+    m_minProbability(0.5f),
+    m_maxNeutrinos(std::numeric_limits<unsigned int>::max()),
+    m_filePathEnvironmentVariable("FW_SEARCH_PATH")
 {
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-void NeutrinoIdTool::SelectOutputPfos(const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, PfoList &selectedPfos)
+void NeutrinoIdTool::SelectOutputPfos(const MasterAlgorithm *const pAlgorithm, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, PfoList &selectedPfos)
 {
     if (nuSliceHypotheses.size() != crSliceHypotheses.size())
         throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
 
+    const unsigned int nSlices(nuSliceHypotheses.size());
+    if (nSlices == 0) return;
+
+    SliceFeaturesVector sliceFeaturesVector;
+    this->GetSliceFeatures(this, nuSliceHypotheses, crSliceHypotheses, sliceFeaturesVector);
+
+    if (m_useTrainingMode)
+    {
+        // ATTN in training mode, just return everything as a cosmic-ray
+        this->SelectAllPfos(crSliceHypotheses, selectedPfos);
+
+        unsigned int bestSliceIndex(std::numeric_limits<unsigned int>::max());
+        if (!this->GetBestSliceIndex(pAlgorithm, nuSliceHypotheses, crSliceHypotheses, bestSliceIndex)) return;
+
+        for (unsigned int sliceIndex = 0; sliceIndex < nSlices; ++sliceIndex)
+        {
+            const SliceFeatures &features(sliceFeaturesVector.at(sliceIndex));
+            if (!features.IsFeatureVectorAvailable()) continue;
+
+            LArSvmHelper::ProduceTrainingExample(m_trainingOutputFile, sliceIndex == bestSliceIndex, features.GetFeatureVector());
+        }
+
+        return;
+    }
+
+    this->SelectPfosByProbability(nuSliceHypotheses, crSliceHypotheses, sliceFeaturesVector, selectedPfos);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void NeutrinoIdTool::GetSliceFeatures(const NeutrinoIdTool *const pTool, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, SliceFeaturesVector &sliceFeaturesVector) const
+{
+    for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
+        sliceFeaturesVector.push_back(SliceFeatures(nuSliceHypotheses.at(sliceIndex), crSliceHypotheses.at(sliceIndex), pTool));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool NeutrinoIdTool::GetBestSliceIndex(const MasterAlgorithm *const pAlgorithm, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, unsigned int &bestSliceIndex) const
+{
+    unsigned int nHitsInBestSlice(0);
```

Comment:
Might want to declare and initialise all three unsigned ints on one line. Have started to find this helpful when reading implementation (conceptually just doing the same thing three times over).